walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,160 @@
+    Collection<TimeRange> meetingOptions = getMeetingRequest(allEventTimes, request.getDuration());
+    Collection<TimeRange> optionalMeetingOptions = new ArrayList<TimeRange>();
+    // get optional attendees open times, if open times overlap with set open times 
+    //  then temp. adjust duration and continue
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> optionalAttendeeTimes = getAllEventTimes(request.getOptionalAttendees(), events);
+      Collections.sort(optionalAttendeeTimes, TimeRange.ORDER_BY_END);
+      optionalMeetingOptions = getMeetingRequest(optionalAttendeeTimes, request.getDuration());
+
+      if (optionalMeetingOptions.isEmpty() && request.getAttendees().isEmpty()) {
+        return Arrays.asList();
+      } else if (allEventTimes.isEmpty()) {
+        return optionalMeetingOptions;
+      }
+
+      // combine event time lists and return optional times if they exist
+      for (TimeRange optional : optionalAttendeeTimes) {
+        allEventTimes.add(optional);
+      }
+
+      Collections.sort(allEventTimes, TimeRange.ORDER_BY_END);
+      optionalMeetingOptions = getMeetingRequest(allEventTimes, request.getDuration());
+      
+      if (!optionalMeetingOptions.isEmpty()) {
+        return optionalMeetingOptions;
+      }
+    }
+
+    return meetingOptions;
   }
+
+  private Collection<TimeRange> getMeetingRequest(ArrayList<TimeRange> eventTimes, Long duration) {
+    Collection<TimeRange> collection = new ArrayList<TimeRange>();
+    int currStart = TimeRange.START_OF_DAY;
+    int currEnd = TimeRange.END_OF_DAY;
+
+    // Create new time which starts at first opening and ends at  of request
+    newTime: 
+    for (TimeRange eventTime : eventTimes) {
+      currEnd = eventTime.start();
+      TimeRange proposeTime = TimeRange.fromStartEnd(currStart, currEnd, false);
+
+      // check that proposed time is longer than request dureation. Also ensures
+      // that the propsed duration is not negtive
+      if (checkDuration(proposeTime, duration)) {
+        // checks if proposed time overlaps other times and makes adjustments
+        // if duration is too short with new time, proposed time is nullified
+        proposeTime = checkProposedTime(proposeTime, duration, eventTimes);
+        if (proposeTime == null) {
+          currStart = eventTime.end();
+          continue newTime;
+        }
+
+        // if the proposed time is already within the collection don't add
+        for (TimeRange requestOption : collection) {
+          if (proposeTime.equals(requestOption)) {
+            currStart = eventTime.end();
+            continue;
+          }
+        }
+
+        collection.add(proposeTime);
+      }
+      currStart = eventTime.end();
+    }
+    //  add time gap for end of day, current start is on last 
+    TimeRange proposeTime = TimeRange.fromStartEnd(currStart, DAY_IN_MINS, false);
+    if (proposeTime.duration() > duration) {
+      collection.add(proposeTime);
+    }
```

Comment:
No need to change your implementation but the current logic has a decent amount of complexity.  You need to perform several iterations over the data multiple times (sort the array, iterate over all the events, iterate over all the proposed times, etc.).  You also had to remember to include the time gap at the end.

Since the minimum time unit is a minute, you could have stored a BitSet where each bit represents whether that minute is free or not.  Then you could loop through all the events once and scratch out possible meeting times.  Then you can iterate over this BitSet once to find the possible meeting blocks.  This adds additional memory but simplifies logic.  Often you'll need to make a decision on whether the tradeoff is worth it.

You could also extend this solution to handle the case of finding a meeting time with the minimum number of conflicts.

Reply:
Ah that makes sense. With the BitSet implementation and adding optional attendees, would you have to keep a 'master' BitSet of the required attendees and then separate BitSets for each optional attendee? I just don't see a clear way to add all optional attendees without a large time complexity and space complexity .